### ISSUE TITLE
[Performance] how I can stop or truncate the generation in onnx gpt-2 model with the beam search option

### ISSUE THREAD

**Issue by Zapotecatl** [microsoft/onnxruntime#13088]
### Describe the issue

I want to stop the generation of gpt2 in order to reduce the time to get the answers, how I can stop or truncate the generation in onnx gpt-2 model with the beam search option?

I run gpt2 in C++ (with CPU) and the answers take on average a time of 8 seconds.

For example:
input: "It was a new experience for me"

output: "_It was a new experience for me, and I couldn't wait to get my hands on it. The first thing you need to do is download the zip file from here: http://www.amazon.com/gp/trade_" 

For the application that I am doing is enoug with the anwser until the dot: "_It was a new experience for me, and I couldn't wait to get my hands on it_." In this way, the calculation after the dot is a inncesary expense.

I have tried to force the value of eos_token_id in the convert_generatoin.py, but it dosen't work.

```
#eos_token_id = config.eos_token_id
eos_token_id = 13 #TODO(Jorge): 13 is the dot in my gpt vocabulary
    pad_token_id = config.eos_token_id if is_gpt2 else config.pad_token_id
    vocab_size = config.vocab_size
```

I appreciate any suggestions, thanks!

### To reproduce

I have exported the GPT 2 model with the Beam Search option with the following command:
`python convert_generation.py -m gpt2 --output C:\\workspace\\onnxruntime\\onnxruntime\\python\\tools\\transformers\\Output2\\gpt2_beam_search.onnx`

### Urgency

I want to publish a demo video game in november 2022.

### Platform

Windows

### OS Version

Windows 10 64 bits

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

ONNX Runtime version: 1.12

### ONNX Runtime API

C++

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

CPU

### Model File

GPT2

### Is this a quantized model?

No

**Comment by GeorgeS2019** [comment]
```mermaid
flowchart TD
    text-->onnx\nmodel-->|com.microsoft.nlp.BeamSearch\nops| output\ntext
```

**Comment by tianleiwu** [comment]
@Zapotecatl, beam search operator will try pad the output to max length when eos token is generated and early stopping is enabled. Did you enable early stopping?

It is not recommended to change EOS token ID after model training.

**Comment by Zapotecatl** [comment]
Hi @tianleiwu , thanks for the answer.

No, I didn't enable early stopping. I looking for that option in the code, then I modified this code line in convert_generation.py:

`model_group.set_defaults(early_stopping=True)`

I have exported my model with this new option in true. I ran the GPT2 (it's actually GPT NEO...) for max_length = 64 and max_length = 128:

**Input**: _Ignacio Perez, an evolutionary biologist from the University of La Paz, and several companions, were exploring the Andes Mountains when they found a small valley, with no other animals or humans._

max_length = 64
**Output**: _' We thought it must be some kind of natural phenomenon,' Perez said._

max_length = 128
**Output**: _' We thoughtit must be some kind of natural phenomenon,' Perez said. It was so quiet that we didn't hear anything until we got close to it.' The group decided to explore further into the valley in order to see if there might be any signs of human habitation. They walked for about two hours before coming across what appeared to be a cave entrance_


As you can see, I continue have the problem. In my application I am using max_lenght = 128 and the idea is stop in the first dot (If there was the option to stop with: ".", "?" And "!", it would be great).

Maybe I'm doing something wrong?

**Comment by Zapotecatl** [comment]
Oh sorry, I had removed the my first eos_code.

The early stopping works very good. The time of the answer is around 1 and 3 seconds.

After, many months, I have achivieved my goal:
+ GPT NEO with with reasonable answers.
+ Answers in practical time only using CPU
+ Onnx in C++ with Unreal engine 4 (I had a lot of problems with pytorch and UE...)

Thank you very much @tianleiwu , I would never have achieved it without your help. Thanks for everything!

**Comment by Zapotecatl** [comment]
I did some tests in "chat mode". I mean something like this:

```
Human: How are you? 
AI: I'm fine. 
Human: What is your name? 
AI: My name is Anna.
```

In this case my solution about force eos_token_id = 13 (dot) doesn't work because GPT NEO avoid write the dot. I return eos_token_id to its standard value ("<|endoftext|>":50256).

The workaround was play with value max_length (I kept the early_stopping = true). The tex generation produce also unnecessary answers, but is is small (I am getting times between 3 and 4 seconds). Something like:

```
AI: I live in London.
Human: where //<---- unnecessary response
```

I think a new option like "stop token" would be fantastic. 

Anyway, the onnx convert_generation.py is amazing!

Regards.
